Fix CPU unit test case

eplanet opened the pull request:
This is the easiest fix I could think of for #218 (and #238). It adds a new setup feature for test cases that will generate the stat file according to the number of CPU available on the system.

Airblader commented on testcases/010-cpu-usage/setup.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Please use 

    #!/usr/bin/env bash

here and below.

Airblader commented on testcases/010-cpu-usage/setup.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+OUTPUT_FILE=$1/stat
```
We should validate / sanitize the argument to avoid attempting to write in / if it isn't set.

Airblader commented:
@stapelberg What do you think about the approach to use a bash script?

stapelberg commented:
> @stapelberg What do you think about the approach to use a bash script?

We don’t currently use shell in i3status at all, so it would be good to implement this in Perl for consistency.

eplanet commented:
Thanks for your comments. I'll update the PR asap. I suppose we should convert [`expected_output.sh`](https://github.com/eplanet/i3status/blob/fix_cpu_tests/travis/run-tests.pl#L22-L23) to Perl as well? It's not used yet anyway.

stapelberg commented:
Yes please.

eplanet commented:
OK guys I updated the PR accordingly. Feel free to add any comment.

stapelberg commented on testcases/010-cpu-usage/setup.pl:
```diff
@@ -0,0 +1,21 @@
+open(my $fh, '>', $output_file) or die "Could not open file '$output_file' $!";
+print $fh "cpu  0 0 0 0 0 0 0 0 0 0\n";
+print $fh "cpu0 100 0 0 0 0 0 0 0 0 0\n";
+print $fh "cpu1 50 0 0 50 0 0 0 0 0 0\n";
```
Are you assuming that all systems have at least two logical cores here? It seems more prudent to only assume one core is present.

eplanet replied:
Those tests were designed for #209 so it didn't make sense to support one core. But now I think we got all we need to support one core as well, I'll update that very soon.

stapelberg commented on travis/run-tests.pl:
```diff
@@ -8,14 +8,19 @@
     my ($dir) = @_;
+
+    if ( -f "@_/setup.pl") {
+        system($^X, "@_/setup.pl", ($dir));
```
Here and below, use the more descriptive `$EXECUTABLE_NAME` instead of `$^X` :). This might require adding `use English;` at the top.

stapelberg commented on travis/run-tests.pl:
```diff
@@ -8,14 +8,19 @@
-    } elsif ( -f "@_/expected_output.sh") {
-        $refres = `bash @_/expected_output.sh`;
+    } elsif ( -f "@_/expected_output.pl") {
+        $refres = capture($^X, "@_/expected_output.pl");
```
What is `capture`? Where is it defined? My `perldoc(1)` doesn’t know about it, and grepping the i3status repo yields no results.

eplanet replied:
You're right, that was an avoidable mistake 😖. It's defined in [Capture/Tiny.pm](http://search.cpan.org/~dagolden/Capture-Tiny-0.46/lib/Capture/Tiny.pm#capture), but using backticks seems to be better anyway.

eplanet commented:
I updated the PR to avoid assuming at least 2 cores, and added a `cleanup.pl` file in test cases to remove generated `stat` files. It must be good to go now 😄.

stapelberg commented on testcases/010-cpu-usage/cleanup.pl:
```diff
@@ -0,0 +1,15 @@
+use warnings;
+
+if ($#ARGV != 0 || not -d $ARGV[0]) {
+    say "Error with setup script: argument not provided or not a directory";
```
s/setup/cleanup/

stapelberg commented on testcases/010-cpu-usage/cleanup.pl:
```diff
@@ -0,0 +1,15 @@
+use strict;
+use warnings;
+
+if ($#ARGV != 0 || not -d $ARGV[0]) {
```
Here and elsewhere: more idiomatic to use `! -d $ARGV[0]` instead of `not -d $ARGV[0]` (the latter differs in precedence from what one would expect, IIRC)

stapelberg commented on testcases/010-cpu-usage/cleanup.pl:
```diff
@@ -0,0 +1,15 @@
+my $output_file = "$ARGV[0]/stat";
+if ( -f $output_file) {
+    unlink $output_file;
+}
```
Here and elsewhere: please add a newline to the end of the file

stapelberg commented on testcases/010-cpu-usage/cleanup.pl:
```diff
@@ -0,0 +1,15 @@
+}
+
+my $output_file = "$ARGV[0]/stat";
+if ( -f $output_file) {
```
nit: `if (-f $output_file)` (remove the whitespace)

stapelberg commented on travis/run-tests.pl:
```diff
@@ -3,19 +3,29 @@
-    } elsif ( -f "@_/expected_output.sh") {
-        $refres = `bash @_/expected_output.sh`;
+    } elsif ( -f "@_/expected_output.pl") {
+        $refres = `/usr/bin/env perl @_/expected_output.pl`;
```
Why not `$EXECUTABLE_NAME @_/expected_output.pl` for consistency with the `system()` calls above and below?

eplanet replied:
This is to catch `stdout` in `$refres` variable. With `system` what you catch is the return code.

stapelberg replied:
Yes. What I was alluding to is your usage of `/usr/bin/env perl` instead of `$EXECUTABLE_NAME`.

eplanet commented:
I took your comments into account. I made my usual typos / mistakes... Sorry about that! I hope this time it's cool. Tested successfully on both multiple cores and solo core.

stapelberg commented:
Thanks!

kousu commented:
This works for me! Thanks for fixing it!

I didn't realize this would be such a convoluted fix. I thought it was just a missing newline. Thanks for tackling it so quickly :)

```
[kousu@phi i3status]$ ./travis/run-tests.pl 
Testing test case '006-cpu-usage-max-threshold-format'… OK
Testing test case '010-cpu-usage'… OK
Testing test case '015-battery-capacity'… OK
Testing test case '008-load-max-threshold-format'… OK
Testing test case '007-cpu-usage-degraded-threshold-format'… OK
i3status: Cannot read CPU usage
Testing test case '012-cpu-usage-error'… OK
Testing test case '011-cpu-usage'… OK
Testing test case '014-battery-capacity'… OK
Testing test case '003-path-exists'… OK
Testing test case '005-cpu-temp-max-threshold-format'… OK
i3status: Cannot read CPU usage
Testing test case '013-cpu-usage-error'… OK
Testing test case '002-cpu-temp'… OK
Testing test case '004-run-watch'… OK
Testing test case '001-battery'… OK
Testing test case '009-disk-low-threshold-format'… OK
[kousu@phi i3status]$ 
[kousu@phi i3status]$ git log --decorate
commit 884e4da6730a31fc7a5ef892fb9cb061baaf5f6c (HEAD -> master, origin/master, origin/HEAD)
Author: Emeric Planet <emeric.planet@datadoghq.com>
Date:   Tue Aug 29 19:01:30 2017 +0200

    Fix CPU unit tests (#239)
    
    Support any amount of available cores on testing machine.
```

eplanet commented:
Great news! Thanks for your feedback. It was indeed not so easy to end up with a nice fix, but that's the price to pay since the code is tightly coupled with system calls... :-P
